src/main/anovos/data_analyzer/quality_checker.py

```diff
@@ -721,185 +713,283 @@ def outlier_detection(
-                    ),
-                )
-                if output_mode == "replace":
-                    odf = odf.drop(i).withColumnRenamed(i + "_outliered", i)
-            else:
-                odf = odf.drop(i + "_outliered")
+                        F.col(i + "_outliered") == -1,
+                        replace_vals[treatment_method][0],
+                    ).otherwise(F.col(i))
+                ),
+            )
+            if output_mode == "replace":
+                odf = odf.drop(i).withColumnRenamed(i + "_outliered", i)
 
-    odf = odf.drop("outliered")
+    def unionAll(dfs):
+        first, *_ = dfs
+        return first.sql_ctx.createDataFrame(
+            first.sql_ctx._sc.union([df.rdd for df in dfs]), first.schema
+        )
+
+    odf_union = unionAll(list_odf)
+    odf_agg = odf_union.groupby("attribute").pivot("outliered").count()
+    odf_print = (
+        odf_agg.withColumn(
+            "lower_outliers", F.col("-1") if "-1" in odf_agg.columns else F.lit(0)
+        )
+        .withColumn(
+            "upper_outliers", F.col("1") if "1" in odf_agg.columns else F.lit(0)
+        )
+        .withColumn("excluded_due_to_skewness", F.lit(0))
+        .select(
+            "attribute", "lower_outliers", "upper_outliers", "excluded_due_to_skewness"
+        )
+        .fillna(0)
+    )
+
+    if skewed_cols:
+        skewed_cols_print = [(i, 0, 0, 1) for i in skewed_cols]
+        skewed_cols_odf_print = spark.createDataFrame(
+            skewed_cols_print, schema=odf_print.columns
+        )
+        odf_print = unionAll([odf_print, skewed_cols_odf_print])
 
     if treatment & (treatment_method == "row_removal"):
-        if skewed_cols:
-            warnings.warn(
-                "Columns dropped from outlier treatment due to highly skewed distribution: "
-                + (",").join(skewed_cols)
-            )
-        for index, i in enumerate(list_of_cols):
-            if i not in skewed_cols:
-                odf = odf.where(
-                    (F.col(i + "_outliered") == 0) | (F.col(i + "_outliered").isNull())
-                ).drop(i + "_outliered")
-            else:
-                odf = odf.drop(i + "_outliered")
+        conditions = [
+            (F.col(i + "_outliered") == 0) | (F.col(i + "_outliered").isNull())
+            for i in list_of_cols
+        ]
+        conditions_combined = functools.reduce(lambda a, b: a & b, conditions)
+        odf = odf.where(conditions_combined).drop(
+            *[i + "_outliered" for i in list_of_cols]
+        )
 
-    if not treatment:
+    if treatment:
+        if output_mode == "replace":
+            odf = odf.select(column_order)
+    else:
         odf = idf
 
-    odf_print = spark.createDataFrame(
-        output_print, schema=["attribute", "lower_outliers", "upper_outliers"]
-    )
     if print_impact:
-        odf_print.show(len(list_of_cols))
+        odf_print.show(len(list_of_cols) + len(skewed_cols), False)
 
     return odf, odf_print
```

Comment:
If possible, can we make it so that it only returns `odf_print` when `print_impact==True`? As we can see in the code, `odf_print` requires a pretty significant calculation power as well, and if the user do not want to see it, there is no reason to go through the calculation everytime.

Reply:
I agree that calculating `odf_print` is relatively time-consuming. My concern is that all functions in `quality_checker` module return two outputs: `odf` and `odf_print`. It might not be suitable if we make `odf_print` optional only for one function. Thus, I have not implemented this change. But in the future we can discuss whether to make `odf_print` optional for all functions in `quality_checker`.

Reply:
As discussed with @kajanansangar and our discussion @sinuochen, `odf_print` should be optional to calculate and will only be returned with `print_impact==True`.